README.md
# Tap Reaction

A flutter plugin made to simplify adding visual feedback when a button is pressed.

There are currently two types of animations availible, `TapReaction.scale()` and `TapReaction.fade()`:

![Two buttons showing a preview of the animations](https://i.imgur.com/DhpH8o9.mp4)

## Usage

To add visual feedback to any widget, simply wrap it in the `TapReaction.scale()` or `TapReaction.fade()` widget. This
widget is fully transparant and won't affect any other widgets that need to recieve the tap (or another gesture).

### Example

```dart
TapReaction.scale(
    endScale: 0.5,
    child: ElevatedButton(
        onPressed: () {
            print('Animating button scale');
        },
        child: Text(
            'Scale',
            style: Theme.of(context).textTheme.headline2,
        ),
    ),
),
```

For a full example see [the example tab](https://pub.dev/packages/tap_reaction/example) or [the GitHub repository](https://github.com/Gothenburg-Application-Studio/tap_reaction/blob/master/example/lib/main.dart).

## Known issues

* The original material design ripple effect is visible even if you use this plugin to get another effect. I have not found a way to disable this for all possible child widgets, and have therefore left that to the user of this plugin. Pull requests are welcome if anyone has a solution.

CHANGELOG.md
## [1.0.0] - 2021-03-14

* Initial release


test/tap_reation_test.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:tap_reaction/tap_reaction.dart';

void main() {
  testWidgets('TapReaction.scale() scales to endScale and back in duration', (WidgetTester tester) async {
    await tester.pumpWidget(
      TapReaction.scale(
        child: Container(
          width: 100,
          height: 100,
        ),
      ),
    );

    Finder _findByWidth(double scale, String description) {
      return find.byWidgetPredicate(
        (Widget widget) {
          if (widget is Transform) {
            Matrix4 transform = widget.transform;
            print(transform.toString());
            return (transform == Matrix4.diagonal3Values(scale, scale, 1.0));
          } else {
            return false;
          }
        },
        description: description,
      );
    }

    expect(
      _findByWidth(1, 'Scale of child widget is 1 at first'),
      findsOneWidget,
    );

    await tester.tap(find.byType(TapReaction));

    // setState after half the animation
    await tester.pump(Duration(milliseconds: 60));

    expect(
      _findByWidth(0.95, 'Scale of child widget is 0.95 after half the duration'),
      findsOneWidget,
    );

    await tester.pumpAndSettle();

    expect(
      _findByWidth(1, 'Scale of child widget is 1 when animation is finished'),
      findsOneWidget,
    );
  });
}


lib/tap_reaction.dart
library tap_reaction;

import 'package:flutter/material.dart';

/// A widget that animates it child when it is pressed.
///
/// To avoid false positives, the animation starts when the pointer is removed
/// from the screen. The same is true for the [InkWell] widget. This means that
/// the animation will start at the same time as an `onTap` in an [InkWell] or
/// an `onPressed` event in an [ElevatedButton].
class TapReaction extends StatefulWidget {
  /// Animates its child using a scale animation.
  ///
  /// More specifically it uses the [Transform.scale()] widget with the `scale`
  /// comes from an animation controller.
  const TapReaction.scale({
    Key? key,
    required this.child,
    this.duration = const Duration(milliseconds: 120),
    this.curve = Curves.easeInCubic,
    double endScale = 0.95,
  })  : animationValue = endScale,
        animationType = AnimationType.scale,
        super(key: key);

  /// Animates its child using a opacity animation.
  ///
  /// More specifically it uses the [Opacity] widget with the `opacity`
  /// comes from an animation controller.
  const TapReaction.fade({
    Key? key,
    required this.child,
    this.duration = const Duration(milliseconds: 120),
    this.curve = Curves.easeInOutCubic,
    double minOpacity = 0.5,
  })  : assert(minOpacity >= 0 && minOpacity <= 1),
        animationValue = minOpacity,
        animationType = AnimationType.fade,
        super(key: key);

  /// The Widget that will react to a tap.
  final Widget child;

  /// Defines how much the widget will be animated.
  ///
  /// This value is specified either via endOpacity or endScale and
  /// must be a value between 1 (no effect) and 0 if [animationType]
  /// is [AnimationType.fade].
  final double animationValue;

  /// The type of animation.
  ///
  /// Eihter `scale` or `fade`. Is specified via the [TapReaction.fade()] or
  /// [TapReaction.scale()] constructor
  final AnimationType animationType;

  // The duration of the animation.
  final Duration duration;

  /// Animation curve
  final Curve curve;

  @override
  _TapReactionState createState() => _TapReactionState();
}

class _TapReactionState extends State<TapReaction>
    with SingleTickerProviderStateMixin {
  late final AnimationController _controller;
  late final Animation<double> _animation;

  /// The treshold for when the pointer event should be seen as a scrolling motion.
  final double _scrollTreshold = 0.2;

  /// Whether the pointer event is a scrolling motion or not
  ///
  /// If the pointer movement is above the `_scrollTreshold` the pointer event
  /// will be seen as a scrolling motion.
  bool _isScroll = false;

  @override
  void initState() {
    _controller = AnimationController(
      vsync: this,
      duration: widget.duration,
    );
    _animation = Tween<double>(begin: 1, end: widget.animationValue).animate(
      CurvedAnimation(
        curve: Curves.linear,
        parent: _controller,
      ),
    )
      ..addListener(() {
        setState(() {});
      })
      ..addStatusListener((status) {
        // Go back to original state.
        if (status == AnimationStatus.completed) _controller.reverse();
      });
    super.initState();
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    // A GestureDetector does not work here since the controller must be
    // animated even if another tap wins in the gesture arena.
    return Listener(
      // Reset variables on a new press.
      onPointerDown: (_) {
        _isScroll = false;
      },
      onPointerMove: (event) {
        // Notify the rest of the widget that this is a scrolling motion.
        if (event.delta.dy.abs() > _scrollTreshold ||
            event.delta.dy.abs() > _scrollTreshold) {
          _isScroll = true;
        }
      },
      onPointerUp: (_) {
        if (!_isScroll) {
          _controller.forward();
        }
      },
      child: widget.animationType == AnimationType.scale
          ? Transform.scale(
              scale: _animation.value,
              child: widget.child,
            )
          : Opacity(
              opacity: _animation.value,
              child: widget.child,
            ),
    );
  }
}

enum AnimationType {
  fade,
  scale,
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:tap_reaction/tap_reaction.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: Scaffold(
        body: Center(
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: [
              TapReaction.scale(
                endScale: 0.5,
                child: ElevatedButton(
                  onPressed: () {
                    print('Animating button scale');
                  },
                  child: Text(
                    'Scale',
                    style: Theme.of(context).textTheme.headline2,
                  ),
                ),
              ),
              SizedBox(
                height: 20,
              ),
              TapReaction.fade(
                minOpacity: 0.2,
                child: ElevatedButton(
                  onPressed: () {
                    print('Animating button opacity');
                  },
                  child: Text(
                    'Fade',
                    style: Theme.of(context).textTheme.headline2,
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
